### PR TITLE
hooks: implement sync phase lifecycle hooks (host and container)

### PR DESCRIPTION
**Pull request by gsquared94**
### Hooks PR 5 of n

Related: #1441, #6147, #6159, #6160, #6163,  #6177
Refer to http://tinyurl.com/skaffold-hooks for detailed design
This PR implements sync stage lifecycle hooks. Try out example `integration/lifecycle-hooks`.  

```
❯ skaffold dev               
.... <build and deploy logs> ....
Deployments stabilized in 20.119678ms
Press Ctrl+C to exit
Watching for changes...
[hooks-example] Hello World!
[hooks-example] Hello World!
Syncing 1 files for hooks-example:bd341ab8cc369fe090a3a833b4ff1b383554d7914ed07b3c255acfdb10c372ce
Starting pre-sync hooks for artifact image "hooks-example:bd341ab8cc369fe090a3a833b4ff1b383554d7914ed07b3c255acfdb10c372ce"...
file changes detected: hello.txt
pre-sync container hook running on hooks-example-pod!
Completed pre-sync hooks for artifact image "hooks-example:bd341ab8cc369fe090a3a833b4ff1b383554d7914ed07b3c255acfdb10c372ce"
Starting post-sync hooks for artifact image "hooks-example:bd341ab8cc369fe090a3a833b4ff1b383554d7914ed07b3c255acfdb10c372ce"...
post-sync container hook running on hooks-example-pod!
Completed post-sync hooks for artifact image "hooks-example:bd341ab8cc369fe090a3a833b4ff1b383554d7914ed07b3c255acfdb10c372ce"
Watching for changes...
[hooks-example] Hello World!!
[hooks-example] Hello World!!
[hooks-example] Hello World!!
^CCleaning up...
 - pod "hooks-example-pod" deleted
```

### CODE REVIEW

File: pkg/skaffold/hooks/container.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+// containerSelector represents a policy for selecting target containers for running a particular lifecycle hook
+type containerSelector func(v1.Pod, v1.Container) (bool, error)
+
+// imageSelector chooses containers that run the given image name
+func imageSelector(image string) containerSelector {
+	return func(p v1.Pod, c v1.Container) (bool, error) {
+		if p.Status.Phase != v1.PodRunning {
+			return false, nil
+		}
+		return c.Image == image, nil
```
Comment (briandealwis) [inline pkg/skaffold/hooks/container.go:43]: You should probably check the corresponding container statuses too (p.Status.ContainerStatuses)?

File: pkg/skaffold/hooks/container.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+	cli        *kubectl.CLI
+	selector   containerSelector
+	namespaces []string
+}
+
+// run executes the lifecycle hook inside the target container
+func (h containerHook) run(ctx context.Context, out io.Writer) error {
+	errs, ctx := errgroup.WithContext(ctx)
+
+	client, err := kubernetesclient.Client()
```
Comment (briandealwis) [inline pkg/skaffold/hooks/container.go:59]: This won't work with deployers with a custom context, right?  Maybe you should pass in the client instance?

File: pkg/skaffold/hooks/sync.go [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+	"path/filepath"
+	"strings"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/kubectl"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/output"
+	v1 "github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest/v1"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+func SyncRunner(cli *kubectl.CLI, image string, namespaces []string, d v1.SyncHooks, opts SyncEnvOpts) Runner {
```
Comment (briandealwis) [inline pkg/skaffold/hooks/sync.go:32]: Convention would new `NewSyncRunner`?

**Comment by tejal29** [comment]
> post-sync container hook running on hooks-example-pod!

Give some quick feedback on output. 
1) Remove image tag from the `Starting pre-sync hooks for artifact image "hooks-example:bd341ab8cc369fe090a3a833b4ff1b383554d7914ed07b3c255acfdb10c372ce"...`
instead just `Starting pre-sync hooks for artifact image "hooks-example"`
2) If a hook fails, is it easy to understand which one failed? If not, can we add a id to the hooks so we can give users information on which hook failed.
```
Starting post-sync hooks for artifact image "hooks-example"
post-sync `id` failed for artifact image "hooks-example"
```

File: pkg/skaffold/hooks/sync.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+		output.Default.Fprintf(out, "Starting %s hooks for artifact image %q...\n", phase, r.image)
+	}
+	env := r.getEnv()
+	for _, h := range hooks {
+		if h.HostHook != nil {
+			hook := hostHook{*h.HostHook, env}
+			if err := hook.run(ctx, out); err != nil {
+				return err
+			}
+		} else if h.ContainerHook != nil {
```
Comment (briandealwis) [inline pkg/skaffold/hooks/sync.go:83]: This needs some additional guards as it's Kubernetes-specific and won't work as-is with the docker deployer

File: pkg/skaffold/hooks/container.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+		for _, p := range pods.Items {
+			for _, c := range p.Spec.Containers {
+				if matched, err := h.selector(p, c); err != nil {
+					return err
+				} else if !matched {
+					continue
+				}
+				args := []string{p.Name, "--namespace", p.Namespace, "-c", c.Name, "--"}
+				args = append(args, h.cfg.Command...)
+				cmd := h.cli.Command(ctx, "exec", args...)
```
Comment (briandealwis) [inline pkg/skaffold/hooks/container.go:79]: Do we configure the kubectl CLI instance with the right context?

File: pkg/skaffold/hooks/container.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	if err != nil {
+		return fmt.Errorf("getting Kubernetes client: %w", err)
+	}
+
+	for _, ns := range h.namespaces {
+		pods, err := client.CoreV1().Pods(ns).List(ctx, metav1.ListOptions{})
+		if err != nil {
+			return fmt.Errorf("getting pods for namespace %q: %w", ns, err)
+		}
+
```
Comment (tejal29) [inline pkg/skaffold/hooks/container.go:69]: I feel like we do this many times. 
e.g in status check, debug container manager. 
Skaffold should not be fetching all pods again and again and should be aware of all containers/pods ?
WDYT?

File: pkg/skaffold/hooks/sync.go (hunk 4 above)
Comment (gsquared94) [reply pkg/skaffold/hooks/sync.go:83]: this is only ever invoked from the `PodSyncer` which is already kubernetes specific. When we have an implementation for the `DockerSyncer` we can create a branch here or have a new file.

File: pkg/skaffold/hooks/container.go (hunk 5 above)
Comment (gsquared94) [reply pkg/skaffold/hooks/container.go:79]: kubectl CLI is configured with the right context. I'm working on a separate PR to fix the `kubernetesclient.Client` references that do not handle kubeContext override during deploy.

File: pkg/skaffold/hooks/container.go (hunk 6 above)
Comment (gsquared94) [reply pkg/skaffold/hooks/container.go:69]: you're right, we do this for `sync`, `deploy`,`status-check`, and a few other places. However caching pod state reliably is difficult IMO. I think there should be a kubernetes event listener that we can subscribe to for doing that, but it's too complicated a change for this PR. I'll create an issue for it.

File: pkg/skaffold/hooks/container.go (hunk 2 above)
Comment (gsquared94) [reply pkg/skaffold/hooks/container.go:59]: you're right.  I'm working on a separate PR to fix the kubernetesclient.Client references that do not handle kubeContext override during deploy.

File: pkg/skaffold/hooks/sync.go (hunk 4 above)
Comment (tejal29) [reply pkg/skaffold/hooks/sync.go:83]: does it makes sense to actually change the SyncHook to also have reference to the containers.  So the containerHook is not k8 specific.

```
func (s *PodSyncer) Sync {
....

containers := fetchAllContainers(kubectl, *s.namespaces)
hooksRunner := hooks.SyncRunner(containers, item.Image, item.Artifact.Sync.LifecycleHooks, opts)
sync()..
containers = fetchAllContainers(kubectl, *s.namespaces)
hooksRunner.updateContainers(containers)
hooksRunner.RunPostHooks(ctx, out)
}
```

That will make sure you don't repeat the pod fetching.

File: pkg/skaffold/hooks/container.go (hunk 1 above)
Comment (gsquared94) [reply pkg/skaffold/hooks/container.go:49]: done.

File: pkg/skaffold/hooks/sync.go (hunk 4 above)
Comment (gsquared94) [reply pkg/skaffold/hooks/sync.go:84]: IDK, it's possible that the container pre-hook crashes the container, or the `sync` step does it. I think it's more reliable with an extra query. Created https://github.com/GoogleContainerTools/skaffold/issues/6233 which should look to address this problem more collectively.

File: pkg/skaffold/hooks/container.go (hunk 6 above)
Comment (gsquared94) [reply pkg/skaffold/hooks/container.go:75]: created https://github.com/GoogleContainerTools/skaffold/issues/6233

**Comment by gsquared94** [comment]
>     1. Remove image tag from the `Starting pre-sync hooks for artifact image "hooks-example:bd341ab8cc369fe090a3a833b4ff1b383554d7914ed07b3c255acfdb10c372ce"...`
>        instead just `Running pre-sync hooks for artifact image "hooks-example"`

Done!

>     2. If a hook fails, is it easy to understand which one failed? If not, can we add a id to the hooks so we can give users information on which hook failed.

With -v DEBUG it'll print the executed command, so it should be easy to tie with the failed hook. I don't want to add an additional `id` property since the command itself should serve as an identifier IMO.

File: pkg/skaffold/hooks/sync.go (hunk 3 above)
Comment (gsquared94) [reply pkg/skaffold/hooks/sync.go]: Done

**Comment by gsquared94** [comment]
currently, the container hooks don't have a podName/containerName prefix like application logs. 
```diff
Starting pre-sync hooks for artifact "hooks-example"...
file changes detected: hello.txt
- [hooks-example] pre-sync container hook running on hooks-example-pod!
+ pre-sync container hook running on hooks-example-pod!
Completed pre-sync hooks for artifact image "hooks-example"
```

Will do this in a follow up PR as it'll need some refactor of the `log` package if I want to retain the same colors as the corresponding application logs. Don't want to bloat this PR.

File: pkg/skaffold/hooks/sync.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+func (r syncRunner) run(ctx context.Context, out io.Writer, hooks []v1.SyncHookItem, phase phase) error {
+	if len(hooks) > 0 {
+		output.Default.Fprintf(out, "Starting %s hooks for artifact %q...\n", phase, r.imageName)
+	}
+	env := r.getEnv()
+	for _, h := range hooks {
+		if h.HostHook != nil {
+			hook := hostHook{*h.HostHook, env}
+			if err := hook.run(ctx, out); err != nil {
+				return err
```
Comment (briandealwis) [inline pkg/skaffold/hooks/sync.go:82]: I think we should log something here:
```
pre-sync hook 4 failed: xxx
```

File: pkg/skaffold/hooks/sync.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+			}
+		} else if h.ContainerHook != nil {
+			hook := containerHook{
+				cfg:        *h.ContainerHook,
+				cli:        r.cli,
+				selector:   imageSelector(r.imageRef),
+				namespaces: r.namespaces,
+			}
+			if err := hook.run(ctx, out); err != nil {
+				return err
```
Comment (briandealwis) [inline pkg/skaffold/hooks/sync.go:92]: This needs some logging too:
```
container hook 1 failed: xxx
```

**Review by briandealwis — APPROVED** [review]
I think we should log hook failures.  Either use the array index or the command executed so the user can determine the failing command?

File: pkg/skaffold/hooks/sync.go (hunk 8 above)
Comment (gsquared94) [reply pkg/skaffold/hooks/sync.go:92]: I thought I fixed these. But my branch push was rejected and I didn't notice 🤦🏽‍♂️

**Review by tejal29 — APPROVED** [review]

**Pull request merged**